### ISSUE TITLE
Allow Webview Panel to be created on its own editor group

### ISSUE THREAD

**Issue by AllanOricil** [microsoft/vscode#125791]
Allow Webview Panel to be Pinned. 

Use case:

As a User, I want my Webviews visible at all times. At the moment this is not possible. When a Webview is visible, and I click on any file in the editor explorer, the webview gets hidden because the file opens in the same editor group where the webview is located. 

Solution:
Add a new config for a Webview Panel that does not allow files to be previewed/opened in that editor group.

**Comment by AllanOricil** [comment]
This is an example of what I want to avoid. 

![Screenshot 2021-06-09 at 20 30 28](https://user-images.githubusercontent.com/55927613/121442247-9abcb080-c961-11eb-97a3-81c43378bd6b.png)

I opened the Webview,  and after that I opened a file, which caused the webview to be hidden, because the file was opened in the same panel. The idea is to create a way to not allow files to be added on the same panel that the webview is displayed.

**Comment by mjbvz** [comment]
I'm updated the issue title since `pinned` means pinned tabs in VS Code. It sounds like what you are asking for is instead an isolated group for the webviews (which isn't a concept that exists today in our editor management). Let me know if that's incorrect

The workaround today is to use a [webview view](https://github.com/microsoft/vscode-extension-samples/tree/main/webview-view-sample)

**Comment by AllanOricil** [comment]
Yes, that is what I mean. The webview view would not be that good because I also want to have the File Explorer Visible. The goal is to keep the Webview visible all the time by never allowing a file to be opened in the same panel/group that webview is, because if this happen the webview is going to be hidden.

**Comment by bpasero** [comment]
I believe this has been implemented for our upcoming release via locked groups: https://github.com/microsoft/vscode-docs/blob/vnext/release-notes/v1_60.md#locked-editor-groups

You can already try this out in insiders today: https://code.visualstudio.com/insiders/

**Comment by AllanOricil** [comment]
@bpasero amazing job!

I will give it a try

**Comment by AllanOricil** [comment]
@bpasero 

I would like to open a new feature to allow developers to set the Webview's panel as a "Locked Group" programatically.

Use Case: The User wants the Webview focused all the time and he does not want files to keep opening in the Webview's panel/group. The User opens and closes the Webview so often that he does not want to always have to manually lock it.

**Comment by bpasero** [comment]
@AllanOricil there is a setting though so that this works exactly like you want. The user has to enable it though: https://code.visualstudio.com/updates/v1_60#_autolock-an-editor-group

**Comment by AllanOricil** [comment]
@bpasero could you provide a simple example where you create a dummy webview that is locked in place automatically?

**Comment by AllanOricil** [comment]
@bpasero I tried to add my webview in the Experimental Auto Lock Groups settings and it did not work. What is the key for a webview that I should add to the Item's column to make it work? I added the name of the panel "Salesforce Query Editor" and it did not work. 

![image](https://user-images.githubusercontent.com/55927613/131998164-d972a7f8-d0b3-4a5b-9fa1-62dc39f7622a.png)

**Comment by bpasero** [comment]
@AllanOricil yeah improving the settings widget to allow you to pick an editor is planned in https://github.com/microsoft/vscode/issues/131587

For now, if you add `workbench.editors.webviewInput` it should auto-lock any kind of editor that shows custom contents. Can you try it? For this feature to work, the editor needs to be opened to the side. Locked groups require more than one group currently.

**Comment by AllanOricil** [comment]
@bpasero setting `workbench.editors.webviewInput` works. I personaly did not like it because it isn't intuitive for the user to set a property called `workbench.editors.webviewInput` and expect that all the Webviews he uses will be locked.  Another argument agains it is the fact that VS Code users are not VS Code Developers, hence they don't know what the term `Webview` means. 

And because of the introduction of this property, I started to think about something else. See if it makes sense. When we talk about Webviews, can you see a scenario where people would want a Unlocked Group? I think everyone that uses a Webview wants it to be visible all the time. Therefore Webviews should always be created in a "Locked Group". Nobody wants an Editor to appear in front of his Webview. I have a feeling that this annoys everyone.  Because they can't actually predict where the Editor will open, unless they remember which "panel" is currently focused. What do you think? If this is true, then I would also use this as an argument to not use this solution for Webviews.

**Comment by bpasero** [comment]
In the future we will allow you to put the exact editor identifier of the webview editor that is specific (e.g. "Markdown Preview"). So what I suggested is just a workaround until https://github.com/microsoft/vscode/issues/131587 is resolved.

> When we talk about Webviews, can you think about a scenario where people would want an Unlocked Group?

Not sure. We did not have locked groups for years, and we had webviews for years and thinking back about feedback I got via GH issues, no-one really asked for this. Your issue here is relatively new.

Maybe users have gotten around this with some usage patterns? I think having the user option to lock a group or not is still good and we can think about enabling it by default for some more webview editors, but I wouldn't enable it for all by default.

**Comment by AllanOricil** [comment]
For everyone who wants to try this workaround, this is how It is done

![image](https://user-images.githubusercontent.com/55927613/132097205-5c22e618-7d6a-41fb-8008-5a4fcdf80f8b.png)

**Comment by stken2050** [comment]
I also developed an extension using Webview Panel that I wanted to have isolated, pinned, and auto-locked. Fortunately, I found a new GUI feature to Lock Group the Webview panel that did exactly what I needed.

![image](https://user-images.githubusercontent.com/1316994/218487429-b0692ded-2e9f-492b-8ad4-19a3e5865e8d.png)

The problem is that users have to do this every time the extension is activated and opened in the Webview panel, so I'm trying to find a way to do it programmatically or with Setttings.

The current one looks like this

![image](https://user-images.githubusercontent.com/1316994/218487858-0e37d895-db2b-4de4-8fd8-358bf3f447f1.png)

and the below does not work.

`workbench.editors.webviewInput` is not allowed in VS Code 1.75.0-insider

```
    "workbench.editor.autoLockGroups": {
        "workbench.editors.webviewInput": true
    }
```

I read #131587 and have no idea to solve this issue. Any idea? @bpasero

**Comment by bpasero** [comment]
Can you give an example how to test your extension, i.e. the source code. Then I can maybe print the identifier of the editor when we consider auto locking. It is possible that maybe the identifier changed or we no longer support that scenario from back then.

**Comment by stken2050** [comment]
> Can you give an example how to test your extension, i.e. the source code. 

Sure.
https://github.com/stken2050/vscode-markdown-note

![image](https://user-images.githubusercontent.com/1316994/218625130-92d80d2b-8173-4448-92fe-74fe3197b6db.png)
It's a WYSIWYG markdown editor in the second pane, and it's required to be isolated for the same reason as the built-in Markdown Preview.

This is based on
https://github.com/microsoft/vscode-webview-ui-toolkit-samples/tree/main/frameworks/hello-world-solidjs

and the exact lines of my WebView creation:
https://github.com/stken2050/vscode-markdown-note/blob/f0a2cc6c95fefe3d26f07e2514061922f3002def/src/panels/NotePanel.ts#L99-L121

`ViewColumn.Two` option lets the WebView open in the second pane.

In an ideal situation, here we also want the **Lock option** within [WebviewPanelOptions](https://code.visualstudio.com/api/references/vscode-api#WebviewPanelOptions)

https://code.visualstudio.com/api/references/vscode-api#window.createWebviewPanel

As [AllanOricil](https://github.com/AllanOricil) commented [on Jun 9, 2021](https://github.com/microsoft/vscode/issues/125791#issue-915627131)  here, the WebView is the only approach for versatile web applications in VSCode and in most of the cases, the apps need to be isolated, and again, we probably want the **Lock option** within [WebviewPanelOptions](https://code.visualstudio.com/api/references/vscode-api#WebviewPanelOptions), not like auto-locking via Settings.

In other words, obviously, the GUI "Lock Group" option is already available and works perfectly. Then all I'm asking here is to check this programmatically from the extension itself, not manually by users.

Thanks for your contribution.

**Comment by bpasero** [comment]
It will work if you add `"mainThreadWebview-showNote": true` to the setting in `workbench.editor.autoLockGroups`.

**Comment by AllanOricil** [comment]
There could exist a method to lock a tab by name programmatically. 
   
                                            T1 stays on top
[[T1][T2]] --> lock(T1) --> [[T1]][[T2]]


New files, or other views, are opened in the first group that isn't locked. 

When opening other views this would happen


T3 opened in the "first" available unlocked group vscode found

 [[T1]][[T2][T3]]

T1 is still visible.

**Comment by stken2050** [comment]
@bpasero  Adding 

```
    "workbench.editor.autoLockGroups": {
        "mainThreadWebview-showNote": true
    }
```
to `settings.json` does work. Thank you.

Actually, I have tried to do this programmatically in multiple ways, but none has worked. So far there is no way to do this from an extension, am I correct?

**Comment by stken2050** [comment]
@AllanOricil I will investigate. Thanks for your advice.

**Comment by bpasero** [comment]
> Actually, I have tried to do this programmatically in multiple ways, but none has worked. So far there is no way to do this from an extension, am I correct?

Maybe @sandy081 can chime in if an extension can write to settings.

**Comment by stken2050** [comment]
eah, I see. understood  🙃

**Comment by KMKoushik** [comment]
@bpasero 
I tried adding "workbench.editors.webviewInput" and "mainThreadWebview-<projectDisplayName>" to settings and it did not work for, Am I missing something?

**Comment by a-stewart** [comment]
@KMKoushik - you want to add the following into your settings.json:

```
  "workbench.editor.autoLockGroups": {
    "mainThreadWebview-myview": true
  }
```

Replacing `myview` with the name the webview uses.

Once you have added it here, you should then also be able to configure it from the Settings UI.

![image](https://github.com/microsoft/vscode/assets/150152/b170ba97-ad8d-407d-9541-0d7588a59720)
